presto-main/src/main/java/com/facebook/presto/execution/buffer/SpoolingOutputBuffer.java

```diff
@@ -0,0 +1,721 @@
+            }
+            pendingRead = this.pendingRead;
+        }
+
+        if (pendingRead != null) {
+            pendingRead.completeResultFutureWithEmpty();
+        }
+    }
+
+    @Override
+    public void fail()
+    {
+        synchronized (this) {
+            // TODO: clean up memory with a close method
+            if (state.setIf(BufferState.FAILED, oldState -> !oldState.isTerminal())) {
+                noMorePages.set(true);
+            }
+        }
+    }
+
+    @Override
+    public long getPeakMemoryUsage()
+    {
+        return 0;
+    }
+
+    @Override
+    public void setNoMorePagesForLifespan(Lifespan lifespan)
+    {
+        // NOOP
+    }
+
+    @Override
+    public void registerLifespanCompletionCallback(Consumer<Lifespan> callback)
+    {
+        // NOOP
+    }
+
+    @Override
+    public boolean isFinishedForLifespan(Lifespan lifespan)
+    {
+        return isFinished();
+    }
+
+    private synchronized void flush()
+    {
+        if (pages.isEmpty()) {
+            return;
+        }
+
+        ImmutableList.Builder<DataOutput> dataOutputs = ImmutableList.builder();
+
+        for (SerializedPage page : pages) {
+            dataOutputs.add(new PageDataOutput(page));
+        }
+
+        // create a future that will hold the handle
+        ListenableFuture<TempStorageHandle> handleFuture = executor.submit(() -> {
+            TempDataSink dataSink = tempStorage.create(context);
+            dataSink.write(dataOutputs.build());
+            return dataSink.commit();
+        });
+
+        // store the handleFuture and file information
+        HandleInfo handleInfo = new HandleInfo(
+                Range.closedOpen(memorySequenceId.get(), memorySequenceId.get() + pages.size()),
+                handleFuture,
+                getPagesSize(pages),
+                pages.size());
+        handleInfoQueue.offer(handleInfo);
+
+        // update cutoff for file pages
+        memorySequenceId.addAndGet(pages.size());
+
+        // clears the pages in memory
+        pages.clear();
+    }
+
+    public void initializeTempStorage()
+            throws IOException
+    {
+        tempStorage.initialize();
+    }
+
+    private List<SerializedPage> getPagesFromStorage(long token, DataSize maxSize, List<HandleInfo> handleInfoList, long startPage)
+            throws CancellationException, ExecutionException, InterruptedException, IOException
+    {
+        synchronized (this) {
+            if (token >= memorySequenceId.get()) {
+                return ImmutableList.of();
+            }
+        }
+
+        Queue<HandleInfo> handleQueue = new ArrayDeque<>(handleInfoList);
+
+        HandleInfo handleInfo = requireNonNull(handleQueue.poll());
+        ListenableFuture<TempStorageHandle> handleFuture = handleInfo.getHandleFuture();
+
+        // get iterator starting at appropriate page
+        InputStream input = closer.register(tempStorage.open(context, handleFuture.get()));
```

Comment:
We shouldn't use the closer to close the input stream--this defers the closing until whenever the closer is closed.  Let's just use try with resources to obtain the stream and close it when we're done with it: https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html